prov/efa/src/rxr/rxr_pkt_type_rts.c

```diff
@@ -0,0 +1,783 @@
+	size_t data_sent;
+
+	tx_entry = (struct rxr_tx_entry *)pkt_entry->x_entry;
+
+	peer = rxr_ep_get_peer(ep, pkt_entry->addr);
+	assert(peer);
+	if (tx_entry->op == ofi_op_read_req) {
+		tx_entry->bytes_sent = 0;
+		tx_entry->state = RXR_TX_WAIT_READ_FINISH;
+		return;
+	}
+
+	data_sent = rxr_get_rts_data_size(ep, rxr_get_rts_hdr(pkt_entry->pkt));
+
+	tx_entry->bytes_sent += data_sent;
+
+	if ((rxr_env.enable_shm_transfer && peer->is_local) ||
+	    !(efa_mr_cache_enable && tx_entry->total_len > data_sent))
+		return;
+
+	/*
+	 * Register the data buffers inline only if the application did not
+	 * provide a descriptor with the tx op
+	 */
+	if (rxr_ep_mr_local(ep) && !tx_entry->desc[0])
+		rxr_inline_mr_reg(rxr_ep_domain(ep), tx_entry);
+}
+
+/*
+ *  rxr_pkt_handle_rts_recv() and related functions
+ */
+char *rxr_pkt_read_rts_base_hdr(struct rxr_ep *ep,
+				struct rxr_rx_entry *rx_entry,
+				struct rxr_pkt_entry *pkt_entry)
+{
+	char *data;
+	struct rxr_rts_hdr *rts_hdr = NULL;
+	/*
+	 * Use the correct header and grab CQ data and data, but ignore the
+	 * source_address since that has been fetched and processed already
+	 */
+
+	rts_hdr = rxr_get_rts_hdr(pkt_entry->pkt);
+
+	rx_entry->addr = pkt_entry->addr;
+	rx_entry->tx_id = rts_hdr->tx_id;
+	rx_entry->msg_id = rts_hdr->msg_id;
+	rx_entry->total_len = rts_hdr->data_len;
+	rx_entry->cq_entry.tag = rts_hdr->tag;
+
+	if (rts_hdr->flags & RXR_REMOTE_CQ_DATA) {
+		rx_entry->cq_entry.flags |= FI_REMOTE_CQ_DATA;
+		data = rxr_get_ctrl_cq_pkt(rts_hdr)->data + rts_hdr->addrlen;
+		rx_entry->cq_entry.data =
+				rxr_get_ctrl_cq_pkt(rts_hdr)->hdr.cq_data;
+	} else {
+		rx_entry->cq_entry.data = 0;
+		data = rxr_get_ctrl_pkt(rts_hdr)->data + rts_hdr->addrlen;
+	}
+
+	return data;
+}
+
+static
+char *rxr_pkt_read_rts_rma_hdr(struct rxr_ep *ep,
+			       struct rxr_rx_entry *rx_entry,
+			       struct rxr_pkt_entry *pkt_entry,
+			       char *rma_hdr)
+{
+	uint32_t rma_access;
+	struct fi_rma_iov *rma_iov = NULL;
+	struct rxr_rts_hdr *rts_hdr;
+	int ret;
+
+	rma_iov = (struct fi_rma_iov *)rma_hdr;
+	rts_hdr = rxr_get_rts_hdr(pkt_entry->pkt);
+	if (rts_hdr->flags & RXR_READ_REQ) {
+		rma_access = FI_SEND;
+		rx_entry->cq_entry.flags |= (FI_RMA | FI_READ);
+	} else {
+		assert(rts_hdr->flags & RXR_WRITE);
+		rma_access = FI_RECV;
+		rx_entry->cq_entry.flags |= (FI_RMA | FI_WRITE);
+	}
+
+	assert(rx_entry->iov_count == 0);
+
+	rx_entry->iov_count = rts_hdr->rma_iov_count;
+	ret = rxr_rma_verified_copy_iov(ep, rma_iov, rts_hdr->rma_iov_count,
+					rma_access, rx_entry->iov);
+	if (ret) {
+		FI_WARN(&rxr_prov, FI_LOG_CQ, "RMA address verify failed!\n");
+		rxr_cq_handle_cq_error(ep, -FI_EIO);
+	}
+
+	rx_entry->cq_entry.len = ofi_total_iov_len(&rx_entry->iov[0],
+						   rx_entry->iov_count);
+	rx_entry->cq_entry.buf = rx_entry->iov[0].iov_base;
+	return rma_hdr + rts_hdr->rma_iov_count * sizeof(struct fi_rma_iov);
+}
+
+static
+int rxr_cq_match_recv(struct dlist_entry *item, const void *arg)
```

Comment:
Minor thing but I'd move these two functions to the top of the file since you have a 'utility functions' block up there.

Reply:
I prefer keep them here because they are only used by rxr_pkt_handle_rts_recv().